src/coreclr/vm/amd64/thunktemplates.asm

```diff
@@ -0,0 +1,80 @@
+; Licensed to the .NET Foundation under one or more agreements.
+; The .NET Foundation licenses this file to you under the MIT license.
+
+include <AsmMacros.inc>
+include AsmConstants.inc
+
+PAGE_SIZE = 4096
+
+DATA_SLOT macro stub, field
+    exitm @CatStr(stub, <Code + PAGE_SIZE + >, stub, <Data__>, field)
+endm
+
+LEAF_ENTRY StubPrecodeCode, _TEXT
+        mov    r10, QWORD PTR [DATA_SLOT(StubPrecode, MethodDesc)]
+        jmp    QWORD PTR [DATA_SLOT(StubPrecode, Target)]
+LEAF_END_MARKED StubPrecodeCode, _TEXT
+
+LEAF_ENTRY FixupPrecodeCode, _TEXT
+        jmp QWORD PTR [DATA_SLOT(FixupPrecode, Target)]
+PATCH_LABEL FixupPrecodeCode_Fixup
+        mov    r10, QWORD PTR [DATA_SLOT(FixupPrecode, MethodDesc)]
+        jmp    QWORD PTR [DATA_SLOT(FixupPrecode, PrecodeFixupThunk)]
+LEAF_END_MARKED FixupPrecodeCode, _TEXT
+
+LEAF_ENTRY CallCountingStubCode, _TEXT
+        mov    rax,QWORD PTR [DATA_SLOT(CallCountingStub, RemainingCallCountCell)]
+        dec    WORD PTR [rax]
+        je     CountReachedZero
+        jmp    QWORD PTR [DATA_SLOT(CallCountingStub, TargetForMethod)]
+    CountReachedZero:
+        call   QWORD PTR [DATA_SLOT(CallCountingStub, TargetForThresholdReached)]
```

Comment:
This has mismatched call/ret count. It is going to work particularly poorly with CET. (I understand that this problem was there even before your change.) 

Can the address of RemainingCallCountCell that is loaded into rax already be used as the stub identifying token instead?

Reply:
That's a good point, it seems there already is a mean to get the stub from the call count cell address.

Reply:
I made it work on x64, I'll implement it that way for all targets.